bpf: run kernel's checkpatch.pl locally and as GitHub action, fix style

qmonnet opened the pull request:
**BACKPORTS**: We want to backport the bpf/ changes (first 5 patches) to avoid conflicts with future backports. However it may not be suitable to backport the new GH action, so **last commit** (`build: run checkpatch.pl for bpf/ locally or as GitHub action`) **should be left out**.

---

Follow-up from RFC #11033.

Although there is no explicit coding style for the BPF code, several developers tend to follow [the kernel coding style conventions](https://www.kernel.org/doc/html/v5.6/process/coding-style.html). To try and maintain a consistent style all over the code, this PR adds kernel's checkpatch.pl to the list of checks performed on the source as part as BPF's `make check`.

Some details on the check:

- We only run checkpatch.pl on the bpf/ directory.

- A new `checkpatch` BPF make target is created, and added as a dependency to `check`.

- However, we only run it on the latest changes by default, meaning the changes since newest Git parent reference plus the changes returned by `git diff HEAD`.

- For manual checks, the bash script used to launch checkpatch has an option to run the checks on all the bpf/ source code, independently of the Git history.

- GitHub action runs checkpatch.pl on all commits from the PR. There is no mechanism to bypass checkpatch errors, Janitors will have to decide if issues reported by checkpatch should be ignored.

The checkpatch.pl and spelling.txt files (both under GPLv2) are directly copied from the kernel repository, as per commit f436a58e2619 (`"checkpatch: fix CONST_STRUCT when const_structs.checkpatch is missing"`) (in linux-next as of the creation of this patch). File `deprecated_terms.txt` is created from this PR, and populated in prevision of the changes proposed for the terminology. The script `checkpatch.sh` (bash script) launches checkpatch.pl with the relevant arguments and options. `COPYING` file is copied from `bpf/COPYING`, to make sure people get a version of the license boilerplate with the checkpatch code.

---

The PR also contains commits to fix a number of style errors and warnings reported by checkpatch.pl. See individual commit logs for details.

I believe fixing them improve code style consistency and readability, but it may complicate backports (although conflicts should be trivial to fix), so as always feedback is welcome.

pchaigno commented on .github/workflows/bpf-checks.yaml:
```diff
@@ -7,6 +7,17 @@ on:
+            - name: Checkout code
+              uses: actions/checkout@v2
+              with:
+                  fetch-depth: 0
```
Did you find this makes a sensible difference? Should we use it in other GitHub workflows (not in this PR of course)?

qmonnet replied:
Lower perf when checking out, but required because I run checkpatch on each individual commit from the PR. From what I understand, without it, checking out the repo is done in such a way it contains only one commit. Other GitHub actions as far as I can tell don't work on the individual commits so no need to change.

pchaigno commented on contrib/checkpatch/checkpatch.pl:
```diff
@@ -0,0 +1,6937 @@
+my $spelling_file = "$D/spelling.txt";
+my $codespell = 0;
+my $codespellfile = "/usr/share/codespell/dictionary.txt";
+my $conststructsfile = "$D/const_structs.checkpatch";
```
Couldn't we remove references to this file in the script instead of having an empty version? Unless you plan to keep this script up-to-date with the kernel's and want to avoid diverging too much? If that's the case, maybe it would be easier to download it directly from the kernel sources in the GitHub action.

qmonnet replied:
Updating the script from the kernel repository from time to time was my intention indeed, so I'd rather avoid patching it if we don't have a strong incentive to do so. Having one empty file under contrib/checkpatch does not seem too expensive?

I thought of downloading the script from the kernel version, but I'd rather keep control of the version we use to avoid bad surprises (e.g. some change regarding a new file needed, that breaks checkpatch if the file is not found and makes all runs of the GitHub action fail all of a sudden.

Regarding the empty file, instead of patching our checkpatch version, I was considering maybe patching upstream so that it doesn't complain if the file is missing but `--ignore CONST_STRUCT` is passed. What do you think? (This specific check is also broken if `CONST_STRUCT` is not ignored but the file is present and empty by the way, so even if we don't really care in our case I could fix that too).

pchaigno replied:
> Regarding the empty file, instead of patching our checkpatch version, I was considering maybe patching upstream so that it doesn't complain if the file is missing but `--ignore CONST_STRUCT` is passed. What do you think? (This specific check is also broken if `CONST_STRUCT` is not ignored but the file is present and empty by the way, so even if we don't really care in our case I could fix that too).

:+1: if you have the time; otherwise LGTM!

qmonnet replied:
It doesn't look too hard (even though it's Perl!). I'll post that soon.

<details><summary>patch</summary>

``` diff
diff --git a/scripts/checkpatch.pl b/scripts/checkpatch.pl
index bf9e0e87a6ef..d91332360494 100755
--- a/scripts/checkpatch.pl
+++ b/scripts/checkpatch.pl
@@ -765,8 +765,10 @@ sub read_words {
 }
 
 my $const_structs = "";
-read_words(\$const_structs, $conststructsfile)
-    or warn "No structs that should be const will be found - file '$conststructsfile': $!\n";
+if (show_type("CONST_STRUCT")) {
+       read_words(\$const_structs, $conststructsfile)
+           or warn "No structs that should be const will be found - file '$conststructsfile': $!\n";
+}
 
 my $typeOtherTypedefs = "";
 if (length($typedefsfile)) {
@@ -6580,7 +6582,8 @@ sub process {
 
 # check for various structs that are normally const (ops, kgdb, device_tree)
 # and avoid what seem like struct definitions 'struct foo {'
-               if ($line !~ /\bconst\b/ &&
+               if ($const_structs ne "" &&
+                   $line !~ /\bconst\b/ &&
                    $line =~ /\bstruct\s+($const_structs)\b(?!\s*\{)/) {
                        WARN("CONST_STRUCT",
                             "struct $1 should normally be const\n" . $herecurr);
```
</details>

[Edit] Now submitted (https://lkml.kernel.org/patchwork/patch/1261008/).
v2: https://lkml.kernel.org/patchwork/patch/1261548/
v3 :astonished:: https://lkml.kernel.org/patchwork/patch/1263025/

pchaigno reviewed (commented):
I'm a bit ashamed of asking this after you modified all files, but is there any benefit in applying all these style changes now? Couldn't we just way for peeps to modify the faulty lines, at which point `checkpatch.pl` will complain and the lines will be fixed? That would be a lot slower, but we wouldn't make backports any bit more difficult (nor would we break `git blame` :grimacing: ).

qmonnet commented on bpf/lib/lb.h:
```diff
@@ -1266,18 +1260,15 @@ static __always_inline int lb4_local(const void *map, struct __ctx_buff *ctx,
-		}
 		backend = lb4_lookup_backend(ctx, slave_svc->backend_id);
-		if (backend == NULL) {
+		if (backend)
```
\*oopsies*

aanm reviewed (approved):
Shouldn't the GH action pass?

joestringer commented on bpf/node_config.h:
```diff
@@ -16,20 +16,23 @@
 #define HOST_IFINDEX_MAC { .addr = { 0xce, 0x72, 0xa7, 0x03, 0x88, 0x56 } }
-#define NAT46_PREFIX { .addr = { 0xbe, 0xef, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0 } }
+#define NAT46_PREFIX { .addr = { 0xbe, 0xef, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, \
+				 0x0, 0x0, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0 } }
```
Short line requirement is annoying here, makes the IP addresses less readable IMO :/

qmonnet replied:
Right, ok I'll revert those. I see some conflicts appeared anyway.

joestringer reviewed (approved):
Not sure if we want to leave a note somewhere in the tree where `checkpatch` came from / how to update it but LGTM.

borkmann reviewed (approved):
lgtm

qmonnet committed:
bpf: fix checkpatch BLOCK_COMMENT_STYLE warnings

Let's harmonise block comment style. The kernel network codestyle is
preferred:

    /* Single-line comment */

    /* Multi-line comment,
     * spanning several lines.
     */

The regular kernel codestyle (non-network-specific) is not reported by
checkpatch, but strongly discouraged, to keep comment blocks consistent:

    /*
     * Multi-line comment,
     * with opening marker on its own line
     */

(Note that we could detect those by tricking checkpatch into believing
the files are under a "net/" directory, but this would also enforce the
"--strict" mode, which we do not desire.)

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
bpf: fix checkpatch warnings

Checkpatch was run as follows:

    $ checkpatch.pl --no-tree --show-types \
        --ignore BLOCK_COMMENT_STYLE --ignore CONST_STRUCT \
        --ignore CONSTANT_CONVERSION --ignore JIFFIES_COMPARISON \
        --ignore MACRO_WITH_FLOW_CONTROL \
        --ignore PRINTK_WITHOUT_KERN_LEVEL --ignore TRAILING_SEMICOLON \
	--ignore VOLATILE \
        -f bpf/**/*.{c,h}

The following warning types were fixed in the bpf/ files:

- ARRAY_SIZE: Prefer ARRAY_SIZE(kernel_hz)
- BRACES: braces {} are not necessary for any arm of this statement
- BRACES: braces {} are not necessary for single statement blocks
- LEADING_SPACE: please, no spaces at the start of a line
- LINE_CONTINUATIONS: Avoid unnecessary line continuations
- LINE_SPACING: Missing a blank line after declarations
- RETURN_VOID: void function return statements are not generally useful
- SPACE_BEFORE_TAB: please, no space before tabs
- SPDX_LICENSE_TAG: Missing or malformed SPDX-License-Identifier tag in
  line 1
- TABSTOP: Statements should start on a tabstop
- UNNECESSARY_ELSE: else is not generally useful after a break or return

The following were ignored:

- BLOCK_COMMENT_STYLE, let's not fix those just yet
- CONST_STRUCT, broken with empty struct list
- CONSTANT_CONVERSION, we do want to use __constant_htons
- CONSTANT_COMPARISON, one false positive in bpf/lib/trace.h, some cases
  in bpf/lib/icmp6.h where we bound the value and it looks better
- JIFFIES_COMPARISON, we do want jiffies
- MACRO_WITH_FLOW_CONTROL, we still want some macros with flow control
- PREFER_ALIGNED, false positive on __aligned definition
- PREFER_PACKED, false postivie on __packed definition
- PRINTK_WITHOUT_KERN_LEVEL, our printk() does not have KERN_<LEVEL>
- TRAILING_SEMICOLON, we do want semicolons in some of our macros
- VOLATILE, we know what we are doing with volatile

The following were ignored, some of which will be fixed in a later
patch. This is because they are downgraded to the "check" level (instead
of warning) when checkpatch runs on files and not on patches and commit
logs (and so I did not see them at first):

- LONG_LINE
- LONG_LINE_COMMENT
- LONG_LINE_STRING
- PREFER_FALLTHROUGH
- TYPO_SPELLING

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
bpf: fix checkpatch errors

Checkpatch was run with the following command:

    checkpatch.pl --no-tree --show-types \
    --ignore COMPLEX_MACRO --ignore MULTISTATEMENT_MACRO_USE_DO_WHILE \
        -f bpf/**/*.{c,h}

The following error types were fixed in the bpf/ sources:

- ASSIGN_IN_IF: do not use assignment in if condition
- CODE_INDENT: code indent should use tabs where possible
- FUNCTION_WITHOUT_ARGS: Bad function definition - xxx() should probably
  be xxx(void)
- INITIALISED_STATIC: do not initialise statics to NULL
- OPEN_BRACE: open brace '{' following struct go on the same line
- OPEN_BRACE: open brace '{' following function definitions go on the
  next line
- POINTER_LOCATION: "(foo*)" should be "(foo *)"
- SPACING: space required before the open brace '{'
- SPACING: space required after that ',' (ctx:VxV)
- SWITCH_CASE_INDENT_LEVEL: switch and case should be at the same indent
- TRAILING_WHITESPACE: trailing whitespace

The following were ignored in this patch:

- COMPLEX_MACRO, because of IPCACHE4_PREFIXES and IPCACHE6_PREFIXES that
  we do not want to enclose between parenthesis.
- MULTISTATEMENT_MACRO_USE_DO_WHILE, we do not want do {} while() for
  macros used to define struct contents.
- SPACING (some of them), when we intentionally aligned fields on
  several lines.
- TRAILING_STATEMENTS, we do not want to add line breaks everywhere in
  bpf/include/bpf/builtins.h.
- Issues reported with lower levels (warnings, minor checks).

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
bpf: fix more checkpatch warnings

Checkpatch was run as follows:

    $ checkpatch.pl --no-tree --strict --show-types \
        --types AVOID_BUG --types FSF_MAILING_ADDRESS \
        --types LONG_LINE --types LONG_LINE_COMMENT \
        --types SPDX_LICENSE_TAG --types TYPO_SPELLING \
        -f bpf/**/*.{c,h}

Although we do not try to fix the reports on the "check" level (logged
when the "--strict" option is used), these are some checks that would
be considered as warnings by checkpatch if not running directly on the
source files (but on patches or git commit instead). Let's process them
to be on par with what the GitHub action would report.

The following warning types were fixed in the bpf/ files:

- LONG_LINE: line length of <n> exceeds 100 columns
- LONG_LINE_COMMENT: line length of <n> exceeds 100 columns
- TYPO_SPELLING: 'xxx' may be misspelled - perhaps 'xxy'?

The following were ignored:

- LONG_LINE_STRING, not desired
- LONG_LINE: IP addresses are better on long lines in bpf/node_config.h
- PREFER_FALLTHROUGH, since "fallthrough;" is not implemented

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
bpf: fix checkpatch C99_COMMENTS errors

Run checkpatch with option "--ignore C99_COMMENT_TOLERANCE" to disable
its default tolerance to C99-style comments (those comments starting
with "// ...").

For this error type, checkpatch.pl also reports the SPDX identifier tags
in .c files as infringing the rule. Those are left untouched.

All other occurrences are replaced with "/* ... */" style comments.

One exception: one comment in bpf/lib/icmp6.h was apparently a leftover
commented instruction and was simply removed.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
build: run checkpatch.pl for bpf/ locally or as GitHub action

Although there is no explicit coding style for the BPF code, several
developers tend to follow the kernel coding style conventions [0]. To
try and maintain a consistent style all over the code, this commit adds
kernel's checkpatch.pl to the list of checks performed on the source as
part as BPF's "make check".

Some details on the check:

- We only run checkpatch.pl on the bpf/ directory.

- A new "checkpatch" BPF make target is created, and added as a
  dependency to "check".

- However, we only run it on the latest changes by default, meaning the
  changes since newest Git parent reference plus the changes returned by
  "git diff HEAD".

- For manual checks, the bash script used to launch checkpatch has an
  option to run the checks on all the bpf/ source code, independently of
  the Git history.

- GitHub action runs checkpatch.pl on all commits from the PR. There is
  no mechanism to bypass checkpatch errors, Janitors will have to decide
  if issues reported by checkpatch should be ignored.

The checkpatch.pl and spelling.txt files (both under GPLv2) are directly
copied from the kernel repository, as per commit f436a58e2619
("checkpatch: fix CONST_STRUCT when const_structs.checkpatch is
missing") (in linux-next as of the creation of this patch). File
deprecated_terms.txt is created from this PR, and populated in prevision
of the changes proposed in #12206. The script checkpatch.sh (bash
script) launches checkpatch.pl with the relevant arguments and options.
COPYING file is copied from bpf/COPYING, to make sure people get a
version of the license boilerplate with the checkpatch code.

[0] https://www.kernel.org/doc/html/v5.6/process/coding-style.html

Signed-off-by: Quentin Monnet <quentin@isovalent.com>
